Realtek: Merge RTL9300 support

bkobl opened the pull request:
This is a series of patches that adds support for booting the RTL9300 SoCs.
It also adds support for IRQ routing on the RTL8390.
The patches allow to boot RTL930X based devices. Run-tested on the Zyxel XGS1210-10 device with an RTL9302B SoC.

matzza commented:
[]()

matzza commented:
@florolf

adschm commented:
Hi, please take care of the formal requirements for commits: https://openwrt.org/submitting-patches

Among them:
- Proper prefix (should be lower-case, so Realtek->realtek)
- Lower case after prefix
- No full stop at the end of the title

adschm commented on target/linux/realtek/dts/rtl930x.dtsi:
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
```
I'd prefer the dts-v1 here in the SoC DTSI as for the other SoC.

adschm commented on target/linux/realtek/dts/rtl930x.dtsi:
```diff
@@ -0,0 +1,132 @@
+		};
+	};
+
+	switch0: switch@bb000000 {
```
Consider using some/any sorting for the nodes.

adschm commented on target/linux/realtek/dts/rtl930x.dtsi:
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
This patch is about adding a timer, so I don't see how adding the full DTSI belongs here.

bkobl commented:
I have completed the patch series with complete support for the RTL9300 SoCs, including a first device using these
drivers (Zyxel XGS1210-12). Sorry for doing this in 2 batches, there was a bug I saw last minute and that took some days to hunt down in the driver. It should be obvious now that the .dtsi actually is for a complete set of network/switc/irq and timer drivers plus a first device.

blogic commented:
@adschm Is it ok if I make sure that the patches are properly annotated ? doing the detail cleanup of this branch of the code makes little sense as we are already pushing large parts of it upstream to mainline. There are 2 parallel efforts going on, there is this version of the code where we bring silicon up based on the SDK code and a completely independent tree where a full re-write of what we see here is happening. I'd propose, that we take this as is and wait for the rewrite to treacle down with v5.10/11

adschm commented:
I will have a closer look later. But generally, I'm not really fond of merging stuff where time hasn't even been enough to spare a few minutes to fix the titles.

adschm commented:
Okay, it appears that this has remained completely untouched since my initial comments.

I don't think that we should just take anything thrown our way when there is no interest to fix at least the most obvious (formal) issues (which are easy and quickly done BTW). There is no benefit in adding this code if it's not done properly, actually it's even worse since people will have to deal with it later ...

Maybe @bkobl can bring this into a mergeable state at some point ...

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+};
+
+&spi0 {
+	status = "okay";
```
Add empty line after status.

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+				label = "jffs2";
+				reg = <0x160000 0x100000>;
+			};
+			partition@b260000 {
```
This is wrong.

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+		phy0: ethernet-phy@0 {
+			reg = <0>;
+			compatible = "ethernet-phy-ieee802.3-c22";
+		};
```
using the macros would be better here ...

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+			phy-is-integrated;
+			reg = <27>;
+		};
+*/
```
Remove the comment.

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+			reg = <0>;
+			label = "lan1";
+			phy-handle = <&phy0>;
+			phy-mode = "internal";
```
macros ...

adschm commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,222 @@
+			phy-mode = "internal";
+			phy-handle = <&phy27>;
+		};
+*/
```
remove comment.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -19,6 +19,15 @@ define Device/Default
+		-e $(if $(KERNEL_ENTRY),$(KERNEL_ENTRY),$(KERNEL_LOADADDR)) \
+		-n '$(1)' -d $@ $@.new
+	mv $@.new $@
+endef
```
This has been dropped in https://github.com/openwrt/openwrt/commit/8582e15616d3d42340810cb9c880825ccf87e1b0

Please use shared Build/uImage instead.

bkobl commented:
@adschm, @blogic, I shall address all the issues raised and resubmit. I am just concerned we will go into that level of detail for all of the 5k lines of code being added. The purpose of this is to make these completely new types of devices (switches that can do routing offloading and provide 10Gig functionality) available for users so that we also learn how to adapt userspace and build a community behind it. We are working very hard to get this mainlined and it will come back in clean form, but this is taking time and so this effort here is about learning how the Linux integration can look like at all and what users can do with it.

blogic commented:
this stuff will treacle down into owrt once @biot got stuff upstream. Please make sure that the patches we shall merge are properly annotated. I am happy to merge them even if they are not perfect. This is our our silicon bringup tree after all ...

Commit:
realtek: add detection of RTL930X SoCs and RTL8313 SoC

This adds support to detect RTL930X based SoCs and the RTL9313 SoC.
Tested on Zyxel XGS1210-10 (RTL9302B SoC) and the
Zyxel XS1930-12 (RTL9313 SoC)

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add support for the RTL8390 and RTL9300 SoC IRQs

This adds support for the RTL8390 and RTL9300 SoCs
it also cleans up unnecessary definitions in mach-rtl83xx.h
and moves definitions relevant for irq routing to irq.h

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add QoS and rate control

This adds support for identifying QoS information in packets
and use this and rate control information to submit to multiple
egress queues. The ethernet driver is also made to support
2 egress and up to 32 egress queues.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: correct l2 offloading tag handling

Makes sure the DSA trailer information on any L2 offloading done
by the switch is honoured by the bridge layer

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: enable default rate limiting and qos settings

Enable default rate limiting and QoS support. Remove
previous storm control code.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: fix RTL8231 gpio expander usage with RTL839X SoC

This fixes the usage of the RTL8231 GPIO extender chip
when used with the RTL839X SoCs. Specifically,
the PHY addresses may be different from 0.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add internal and external SDS/PHYs of RTL9300 devices

This adds support for the internal SerDes of the RTL9300 SoC
and for the RTL8218D and RTL8226B phys found in combination
with this SoC in switches.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add switch driver support for the RTL93XX based switches

Adds support for the RTL9300 and RTL9310 series of switches
with 10GBit per port and up to 56 ports.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add API for the hw tables of RTL83XX/93XX SoCs

Add a table API that has per accss register locking and uses
register description information to handle all table access
through a single set of api calls.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add debugfs support for mirroring

This adds debugfs support to monitor mirroring via debugfs

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: new device support ZyXEL XGS1210-12

The ZyXEL XGS1210-12 is a 8x1GBit, 2x2.5GBit and 2x10GBit SFP+
switch. The device has the following hardware:
    - RTL9302B SoC
    - Macronix MX25L12833F (16MB flash)
    - Nanja NT5CC64M16GP-1 (128MB DDR3 SDRAM)
    - RTL8231 GPIO extender to control the port LEDs
    - RTL8218D 8x Gigabit PHY
    - 2 RTL8226 2.5 Gigabit PHYs
The 2 SFP+ uplink ports are not supported by the realtek switch driver, yet.
The bi-color and tri-color (for the 2.5 Gig ports) LEDs are driven directly by
the SoC. A reset button is present on the front but not supported.
Behind the air-vents on the right a 3.3V uart header is externally accessible.
Pins are from top to bottom Vcc(3.3V), TX, RX and GND. Serial connection is via
115200 baud.

To install, boot into u-boot, halt via pressing <ESC>, load initramfs via tftp
and boot via bootm.
Installation to flash was not tested.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

adschm commented:
I still see the entire SoC DTSI added in a patch that tells me it's doing something with the timer. That should definitely be separate ...

adschm commented:
Looks okay formally from a (very) quick look now, apart from the DTSI added in the third patch (please split that) and the custom-uimage function in the last commit that still needs to be removed.

Thanks for updating this so quickly.

bkobl commented:
The .dtsi is in the 3rd patch because this is the first in the set of patches that actually allows to boot an RTL9300 based device up to the prompt, the previous ones will just be able to detect the SoC (1st one), then be able to use serial lines (the IRQs are there) but without the timer this device does nothing as the internal MIPS R4K timer is disabled (at least its IRQ, which is misused as the external IRQ for exactly this external SoC timer). So this third patch actually allows to boot a device for the first time to the prompt, but _only_ if you have the .dtsi with the timer configuration. For me it feels this is exactly the right point in the sequence to add the .dtsi, because up to that point an RTL838X-dtsi would also have worked up to the point where the timer is configured, only.

adschm commented:
> The .dtsi is in the 3rd patch because this is the first in the set of patches that actually allows to boot an RTL9300 based device up to the prompt, the previous ones will just be able to detect the SoC (1st one), then be able to use serial lines (2nd: the IRQs are there) but without the timer this device does nothing as the internal MIPS R4K timer is disabled (at least its IRQ, which is misused as the external IRQ for exactly this external SoC timer). So this third patch actually allows to boot a device for the first time to the prompt, but _only_ if you have the .dtsi with the timer configuration. For me it feels this is exactly the right point in the sequence to add the .dtsi, because up to that point an RTL838X-dtsi would also have worked up to the point where the timer is configured, and the device gets stuck.

And what I'm looking for is simply adding the dtsi in a separate fourth commit after this third commit: Third commits adds working timer support. Now we have all the prerequisites, so we can add the DTSI. But this is a separate step. I'm simply looking for this separation, so adding the DTSI is not "hidden" inside a commit with a different issue. Simply splitting this out with a two sentence commit message should do the job and still be a quick/simple task.

Commit:
realtek: add support for the RTL9300 timer

this adds support for the SoC timer of the RTL9300 chips, it
provides 6 independent timer/counters, of which the first one
is used as a clocksource and the second one as event timer.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add dtsi for RTL930X SoCs

Add a default dtsi to support RTL930X SoCs.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

bkobl commented:
The commit has been split into the timer and the dtsi.

adschm commented:
@blogic I have no objections anymore if you want to merge this now. Note that I was discussing formal stuff here, so you need to be sure that the content is appropriate as well ;-)

Note that somebody, either you or @bkobl , still needs to remove the Build/custom-uimage block in the last commit.

bkobl commented:
Done. Someone explained to me how to use git --amend and now things are easier...

mkresin commented on target/linux/realtek/dts/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,121 @@
+			partition@300000 {
+				label = "firmware";
+				reg = <0x300000 0xd00000>;
+				compatible = "denx,uimage";
```
> Installation to flash was not tested.

Not sure about the reason. But since Bjørns "[use device tree properties for non-standard uimage parsing](https://patchwork.ozlabs.org/project/openwrt/list/?series=225499&state=*)" patches are merged, the image splitting should work with the following changes:

```
compatible = "openwrt,uimage", "denx,uimage";
openwrt,partition-magic = <0x93001210>;
```

bkobl replied:
Flashing was not tested because I was not sure to get the flash back into factory configuration. I also want to discourage people who believe this device will already flawlessly run. In contrast to other Realtek switches, u-boot on this device does not properly initialize and patch the 2.5 GBit PHYs and the 10GBit SerDes', the SDK code suggests without proper patching there might be compatibility issues in particular with Marvel chips. For efficient testing I will need to go back and forth for some more time between the original firmware and owrt. I would be happy if someone else wants to add the compatibility properties and try flashing.

adschm replied:
I'd be happy if we not merge this device support commit then, and keep it open as a separate PR after the rest has been merged. The normal expection is that stuff which is build by buildbots can be expected to work.

blogic replied:
I agree, the rest should be good to go i think, I just gave it a spin locally and things appear to be in place

ynezz commented:
Thanks, merged it without that new device into my staging tree, will merge later once the build is finished.

ynezz commented:
@bkobl 

FYI, following config:

```shell
CONFIG_TARGET_realtek=y
CONFIG_TARGET_realtek_generic=y
CONFIG_TARGET_realtek_generic_DEVICE_d-link_dgs-1210-28=y
CONFIG_DEVEL=y
```

provides following compile error:

```shell
make[7]: *** No rule to make target 'drivers/clocksource/timer-rtl9300.o', needed by 'drivers/clocksource/built-in.a'.  Stop.
scripts/Makefile.build:496: recipe for target 'drivers/clocksource' failed
```

bkobl commented:
Thanks for letting me know, I am investigating.

bkobl commented:
Added forgotten target/linux/realtek/files-5.4/drivers/clocksource/timer-rtl9300.c and rebased, sorry @ynezz.

ynezz commented:
Thanks for the quick fix.

Neustradamus commented:
Merged commits:
- https://github.com/openwrt/openwrt/commit/cb39602399e3c3ee48b4f2c39df7bf759418584e
- https://github.com/openwrt/openwrt/commit/1b1bb6bf19ff1ae4d31910f2326404597a7e8470
- https://github.com/openwrt/openwrt/commit/6f67c079e61a01d599af34dbb852a4fe99d0f52d
- https://github.com/openwrt/openwrt/commit/27029277f98ddd0006175bdc5beec8b0b150f187
- https://github.com/openwrt/openwrt/commit/d140b9fc23f6b58ce6be696844e216c91e62b4c5
- https://github.com/openwrt/openwrt/commit/e1bca7c77d80b3f35f5045cd8983349819165f1e
- https://github.com/openwrt/openwrt/commit/6aac20bc0187e0934a9c167f21960202be182546
- https://github.com/openwrt/openwrt/commit/56ac4fced086c74c73cc05a6af750ee236910811
- https://github.com/openwrt/openwrt/commit/dc9cc0d3e2a1ede4bcf2e6b5b0aded4dc50244af
- https://github.com/openwrt/openwrt/commit/4237488cf3de13b56484f728b77bdc746234954b
- https://github.com/openwrt/openwrt/commit/8faffa00cb6b7121fbdac3d5cff8bab9a253ae93
- https://github.com/openwrt/openwrt/commit/03c79ca269e612faa55e3decd034c44f9b77c05c
- https://github.com/openwrt/openwrt/commit/cb39602399e3c3ee48b4f2c39df7bf759418584e
